### ISSUE TITLE
Some exceptions on startup don't get logged due to AsyncAppender

### ISSUE THREAD

**Issue by hdeadman** [apereo/cas#2368]
If I startup CAS from maven and get a fatal error (e.g. spring-cloud-zookeeper can't resolve host - but I have seen it previously for other reasons), the exception isn't logged and output looks like this:

```
  __  ____     _     ____  __
 / / / ___|   / \   / ___| \ \
| | | |      / _ \  \___ \  | |
| | | |___  / ___ \  ___) | | |
| |  \____|/_/   \_\|____/  | |
 \_\                       /_/

CAS Version: 5.0.2
Apache Tomcat Version: Apache Tomcat/8.5.9
Build Date/Time: 2017-02-01T18:24:32.687Z
System Temp Directory: C:\Users\xyz\AppData\Local\Temp\
Java Home: C:\Program Files\Java\jdk1.8.0\jre
Java Vendor: Oracle Corporation
Java Version: 1.8.0_111
JCE Installed: no
OS Architecture: amd64
OS Name: Windows 7
OS Version: 6.1

←[0m
2017-02-02 16:20:04,791 WARN [org.springframework.boot.SpringApplication] - <Err
or handling failed (ApplicationEventMulticaster not initialized - call 'refresh'
 before multicasting events via the context: org.springframework.boot.context.em
bedded.AnnotationConfigEmbeddedWebApplicationContext@25a65b77: startup date [Wed
 Dec 31 19:00:00 EST 1969]; parent: org.springframework.context.annotation.Annot
ationConfigApplicationContext@66048bfd)>
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:13 min
[INFO] Finished at: 2017-02-02T16:20:04-05:00
[INFO] Final Memory: 12M/240M
[INFO] ------------------------------------------------------------------------
```
If I change all the <AsyncLogger in log4j2.xml to <Logger then I get the actual error:
```
CAS Version: 5.0.2
Apache Tomcat Version: Apache Tomcat/8.5.9
Build Date/Time: 2017-02-01T18:24:32.687Z
System Temp Directory: C:\Users\xyz\AppData\Local\Temp\
Java Home: C:\Program Files\Java\jdk1.8.0\jre
Java Vendor: Oracle Corporation
Java Version: 1.8.0_111
JCE Installed: no
OS Architecture: amd64
OS Name: Windows 7
OS Version: 6.1

←[0m
2017-02-02 16:22:45,318 WARN [org.springframework.boot.SpringApplication] - <Err
or handling failed (ApplicationEventMulticaster not initialized - call 'refresh'
 before multicasting events via the context: org.springframework.boot.context.em
bedded.AnnotationConfigEmbeddedWebApplicationContext@2ed0fbae: startup date [Wed
 Dec 31 19:00:00 EST 1969]; parent: org.springframework.context.annotation.Annot
ationConfigApplicationContext@445b84c0)>
2017-02-02 16:22:45,345 ERROR [org.springframework.boot.SpringApplication] - <Ap
plication startup failed>
java.lang.reflect.UndeclaredThrowableException
        at org.springframework.util.ReflectionUtils.rethrowRuntimeException(ReflectionUtils.java:317) ~[spring-core-4.3.4.RELEASE.jar!/:4.3.4.RELEASE]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySource.findProperties(ZookeeperPropertySource.java:104) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySource.<init>(ZookeeperPropertySource.java:46) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySourceLocator.create(ZookeeperPropertySourceLocator.java:116) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySourceLocator.locate(ZookeeperPropertySourceLocator.java:101) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        at org.springframework.cloud.bootstrap.config.PropertySourceBootstrapConfiguration.initialize(PropertySourceBootstrapConfiguration.java:89) ~[spring-cloud-context-1.1.6.RELEASE.jar!/:1.1.6.RELEASE]
        at org.springframework.boot.SpringApplication.applyInitializers(SpringApplication.java:635) [spring-boot-1.4.2.RELEASE.jar!/:1.4.2.RELEASE]
        at org.springframework.boot.SpringApplication.prepareContext(SpringApplication.java:349) [spring-boot-1.4.2.RELEASE.jar!/:1.4.2.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:313) [spring-boot-1.4.2.RELEASE.jar!/:1.4.2.RELEASE]
        at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:134) [spring-boot-1.4.2.RELEASE.jar!/:1.4.2.RELEASE]
        at org.apereo.cas.web.CasWebApplication.main(CasWebApplication.java:61)
[cas-server-webapp-init-5.0.2.jar!/:5.0.2]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_111]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:1.8.0_111]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_111]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_111]
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [cas-pki-war-1.0.war:?]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [cas-pki-war-1.0.war:?]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:50) [cas-pki-war-1.0.war:?]
        at org.springframework.boot.loader.WarLauncher.main(WarLauncher.java:59)
 [cas-pki-war-1.0.war:?]
Caused by: java.net.UnknownHostException: zookeeper1
        at java.net.Inet4AddressImpl.lookupAllHostAddr(Native Method) ~[?:1.8.0_111]
        at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:928) ~[?:1.8.0_111]
        at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1323) ~[?:1.8.0_111]
        at java.net.InetAddress.getAllByName0(InetAddress.java:1276) ~[?:1.8.0_111]
        at java.net.InetAddress.getAllByName(InetAddress.java:1192) ~[?:1.8.0_111]
        at java.net.InetAddress.getAllByName(InetAddress.java:1126) ~[?:1.8.0_111]
        at org.apache.zookeeper.client.StaticHostProvider.<init>(StaticHostProvider.java:61) ~[zookeeper-3.4.6.jar!/:3.4.6-1569965]
        at org.apache.zookeeper.ZooKeeper.<init>(ZooKeeper.java:445) ~[zookeeper-3.4.6.jar!/:3.4.6-1569965]
        at org.apache.curator.utils.DefaultZookeeperFactory.newZooKeeper(DefaultZookeeperFactory.java:29) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.CuratorFrameworkImpl$2.newZooKeeper(CuratorFrameworkImpl.java:150) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.HandleHolder$1.getZooKeeper(HandleHolder.java:94)
~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.HandleHolder.internalClose(HandleHolder.java:128)
~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.HandleHolder.closeAndReset(HandleHolder.java:77) ~
[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.ConnectionState.reset(ConnectionState.java:217) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.ConnectionState.checkTimeouts(ConnectionState.java:193) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.ConnectionState.getZooKeeper(ConnectionState.java:88) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.CuratorZookeeperClient.getZooKeeper(CuratorZookeeperClient.java:116) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.CuratorFrameworkImpl.getZooKeeper(CuratorFrameworkImpl.java:489) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.GetChildrenBuilderImpl$3.call(GetChildrenBuilderImpl.java:214) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.GetChildrenBuilderImpl$3.call(GetChildrenBuilderImpl.java:203) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.RetryLoop.callWithRetry(RetryLoop.java:108) ~[curator-client-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.GetChildrenBuilderImpl.pathInForeground(GetChildrenBuilderImpl.java:200) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.GetChildrenBuilderImpl.forPath(GetChildrenBuilderImpl.java:191) ~[curator-framework-2.10.0.jar!/:?]
        at org.apache.curator.framework.imps.GetChildrenBuilderImpl.forPath(GetChildrenBuilderImpl.java:38) ~[curator-framework-2.10.0.jar!/:?]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySource.getChildren(ZookeeperPropertySource.java:116) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        at org.springframework.cloud.zookeeper.config.ZookeeperPropertySource.findProperties(ZookeeperPropertySource.java:81) ~[spring-cloud-zookeeper-config-1.0.3.RELEASE.jar!/:1.0.3.RELEASE]
        ... 17 more
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:12 min
[INFO] Finished at: 2017-02-02T16:22:45-05:00
[INFO] Final Memory: 12M/240M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.springframework.boot:spring-boot-maven-plugin
:1.4.2.RELEASE:run (default-cli) on project cas-pki-config: Could not exec java:
 Application finished with exit code: 1 -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

This particular error is my fault but when I first started using CAS I was bit by this a few times for different errors. Sometimes I would get lucky and see the error (timing?) or I would see it if I used the debugger. Eventually I just made my own log4j2.xml. 

I haven't tried this but something like this in CasWebApplication might fix the issue:
```
              try {
			new SpringApplicationBuilder(CasWebApplication.class).banner(new CasBanner())
					.properties(Collections.singletonMap(
							CasEmbeddedContainerConfiguration.EMBEDDED_CONTAINER_CONFIG_ACTIVE, Boolean.TRUE))
					.logStartupInfo(true).run(args);
		} finally {
			org.apache.logging.log4j.LogManager.shutdown();
		}
```
That is referencing log4j2 directly instead of slf4j so maybe that isn't ideal. Other options might be to just sleep for a second in the finally block or not use AsyncAppender.

**Comment by mmoayyed** [comment]
Have you tried raising this issue with log4j folks or even spring boot?

**Comment by mmoayyed** [comment]
Is this also something that can be duplicated with 5.1? Given it runs a more recent version of log4j.

**Comment by mmoayyed** [comment]
Separately and out of interest, what are your intentions with zookeeper?

**Comment by hdeadman** [comment]
I can try it out in 5.1, there is that log4j servlet context listener bean that should be shutting down logging if spring-boot is shutting down the servlet context and all the spring beans. Not sure if it flushes all the async loggers... 

Regarding zookeeper, we have Solr Cloud used by some applications and it relies on zookeeper. Since we have zookeeper in the environment we are using it as key value store to provide an option for overriding some properties in an environment without packaging them with the webapp or the app server container. Something like Vault (which I haven't used) might be better than zookeeper for what we are doing but we already have zookeeper... 

Solr uses zookeeper to store whole config files but unless I am missing something, spring-cloud-zookeeper uses the "file" you put in zookeeper as the "key" and the contents of the file are the value. I can use ${someVariable} in a cas property file and then spring-cloud-zookeeper will look in zookeeper for various paths (e.g. /config/application, /config/cas, /config/cas,somprofile, etc) and in each of those paths it will look for "files" and if one of those is "someVariable" then it will use the contents as the value for that property. I was surprised it didn't look for property files and load them instead of single properties.

**Comment by hdeadman** [comment]
I am trying to try this out with 5.1.0-RC1 but when I switch to that I am getting a conflict on port 8443. The port is not in use prior to me starting spring-boot:run. Is there anything new regarding the tomcat configuration? I also got the same error with 5.1.0-SNAPSHOT (and for that I upgraded spring boot to 1.5.1). 

```
***************************
APPLICATION FAILED TO START
***************************

Description:

The Tomcat connector configured to listen on port 8443 failed to start. The port may already be in use or the connector may be misconfigured.

Action:

Verify the connector's configuration, identify and stop any process that's listening on port 8443, or configure this application to listen on another port.
```

**Comment by mmoayyed** [comment]
There isnt anything new as far as I remember. Are you running embedded?

**Comment by hdeadman** [comment]
It wasn't a port conflict issue, it was a mis-configuration but I wasn't seeing the error logged, I think because org.apache log level is set to off, and that includes org.apache.tomcat.* so an error level log message was not logged. Is there ever a reason to turn a log level off? When wouldn't you want to see errors logged?

        <AsyncLogger name="org.apache" level="off" additivity="false">
            <AppenderRef ref="casConsole"/>
            <AppenderRef ref="casFile"/>
        </AsyncLogger>

**Comment by mmoayyed** [comment]
Cool. This is likely an oversight. I think the intention here was to turn off warnings really that tomcat produces when it scans for tlds and jars and somehow the level got set to off. If you'd rather reverse that please submit a PR.

**Comment by mmoayyed** [comment]
I am also assuming the logging change addresses the original issue as well?

**Comment by hdeadman** [comment]
I was able to get a stack trace in 5.0.2 when I changed the Loggers from async to regular without changing any log levels. The same stacktrace wasn't logged when the loggers were async. I still haven't re-created that situation in 5.1 but only because I ran into some other issues when trying to get 5.1 going that I eventually got around (there was a mix of groovy versions that groovy didn't like, and the newer guava version didn't seem to be compatible with the zookeeper/curator jars I was using so I had to use an older version.

**Comment by mmoayyed** [comment]
OK, cool. Keep updating the 5.1 snapshot and continue to try to duplicate the issue. Log4j JIRA shows that a number of similar and related issues in this area got fixed in log4j 2.7.

Also consider: https://apereo.github.io/cas/development/installation/Logging.html#asyncloggers-shutdown-with-tomcat

**Comment by mmoayyed** [comment]
Going to close this for now, in favor of the new changes/fixes available in master and current `SNAPSHOT` releases. If you do run into this issue again, please post back so we can reopen the issue and investigate.
